lib/kernel.js

```diff
@@ -16,25 +16,107 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import type {
+  HydrogenKernelMiddlewareStack,
+  HydrogenKernelMiddleware
+} from "./plugin-api/hydrogen-types";
 import InputView from "./input-view";
 import KernelTransport from "./kernel-transport";
 import type { ResultsCallback } from "./kernel-transport";
 
-interface KernelMiddlewareStack {
-  interrupt: () => void;
-  shutdown: () => void;
-  restart: (onRestarted: ?Function) => void;
-  execute: (
+// Adapts middleware objects provided by plugins to an internal interface. In
+// particular, this implements fallthrough logic for when a plugin defines some
+// methods (e.g. execute) but doesn't implement others (e.g. interrupt). Note
+// that HydrogenKernelMiddleware objects are mutable: they may lose/gain methods
+// at any time, including in the middle of processing a request.
+// TODO(nikita): insert checks between plugins that ensure that messages being
+// passed around via results callbacks comply with the Jupyter messaging spec.
+class MiddlewareAdapter implements HydrogenKernelMiddlewareStack {
+  _middleware: HydrogenKernelMiddleware;
+  _next: MiddlewareAdapter | KernelTransport;
+  constructor(
+    middleware: HydrogenKernelMiddleware,
+    next: MiddlewareAdapter | KernelTransport
+  ) {
+    this._middleware = middleware;
+    this._next = next;
+  }
+
+  // The return value of this method gets passed to plugins! For now we just
+  // return the MiddlewareAdapter object itself, which is why all private
+  // functionality is prefixed with _, and why MiddlewareAdapter is marked as
+  // implementing HydrogenKernelMiddlewareStack. Something more complicated may
+  // to handle plugin API compatibility, once multiple API versions exist.
+  get _nextAsPluginType(): HydrogenKernelMiddlewareStack {
```

Comment:
Again, there is no need for plugins to receive the next plugin. This should be handled by Hydrogen (by `MiddlewareAdapter`).

To be clear, there is no need for `HydrogenKernelMiddlewareStack`. Plugins should implement `HydrogenKernelMiddleware`. Plugins should invoke the next plugin directly. The callback passed to a plugin should be the one handling the invocation of the next plugin.

---

I don't see any pros to passing the next plugin directly. Only cons; e.g. if we want to mix plugins of different versions, we'd have to create adapters between different version.

Reply:
> there is no need for plugins to receive the next plugin

> Plugins should invoke the next plugin directly.

The above two statements are contradictory.

---

I don't understand what you're trying to say. Can you provide a concrete example (preferably using code). Here are two sample plugins that you could consider as coexisting at the same time:

Plugin 1: duplicates all execute requests
```js
{execute: (next, code, callWatches,onResults) => {
  next.execute(code, callWatches, (message, channel) => {
    onResults(message, channel);
    if (isLastMessage(message)) {
      next.execute(code, callWatches, onResults);
    }
  }
}}
```

Plugin 2: modifies inspect requests
```js
{inspect: (next, code, cursorPos, onResults) => {
  next.inspect(modifyCode(code), cursorPos, onResults);
}}
```

Reply:
> > there is no need for plugins to receive the next plugin#
> >
> > Plugins should invoke the next plugin directly.
>
> The above two statements are contradictory.

It's a typo, it should've read *"Plugins shouldn't invoke the next plugin directly."*

Reply:
Ok, glad there's no contradiction! Point about not understanding what you're saying still stands.

Reply:
Note that when I wrote that comment I was still under the belief that `next` was an instance of a plugin. Since the last patchset introduced `MiddlewareAdapter`, `next` is actually an instance `MiddlewareAdapter`

This is ironic, because I now see that instead of `HydrogenKernelMiddleware`, the only interface we need is `HydrogenKernelMiddlewareStack`. Note that you can achieve this simply by changing `MiddlewareAdapter`'s constructor to:

```js
  constructor(
    middleware: HydrogenKernelMiddlewareStack,
    next: MiddlewareAdapter | KernelTransport
  ) {
    this._middleware = middleware;
    this._next = next;
}
```

I guess this is the best chance we have to get rid of the name `HydrogenKernelMiddlewareStack`. We can remove `HydrogenKernelMiddleware`, and rename `HydrogenKernelMiddlewareStack` to `HydrogenKernelMiddleware`.

---
One more thing. As I said earlier, I think plugin 1 must invoke `onResults` only once; i.e.:
```js
{execute: (next, code, callWatches,onResults) => {
  next.execute(code, callWatches, (message, channel) => {
    next.execute(code, callWatches, (message, channel) => {
        onResults(message, channel);
    });
  });
}}
```

Reply:
The code
```js
constructor(
    middleware: HydrogenKernelMiddlewareStack,
    next: MiddlewareAdapter | KernelTransport
  ) {
    this._middleware = middleware;
    this._next = next;
}
```

doesn't work because now
```js
this._middleware.interrupt(this._nextAsPluginType);
```
is invalid (since `this._middleware` is an instance of `HydrogenKernelMiddlewareStack`, and `HydrogenKernelMiddlewareStack.execute` does not take `next` as its first argument).

You have to do something like options B or C in my previous post to address this.

Reply:
@nikitakit I got the names wrong, I mean we need to keep the interface that takes `next` as one of the arguments.

Reply:
`next.execute` can't take `next` as it's first argument -- when I call `next.execute` how do I know what the next-next plug in is?

Reply:
> when I call next.execute how do I know what the next-next plug in is?

Since `next` itself is an instance of `MiddlewareAdapter`, then the next-next plugin is reachable via `next._next`.

Reply:
So you're saying that we have every plugin do `next.execute (next.next, ...)` every time it wants to call something?

And for the plug in to have the option of skipping it's children (since now `next.next` is public API)?

I think it's worth the extra code complexity in hydrogen to save plugins the hassle and plugin-side complexity of doing this. And to not open the possibility of plug-ins abusing this, e.g. by manually walking the `next` chain and modifying it.

Reply:
> So you're saying that we have every plugin do next.execute (next.next, ...) every time it wants to call something?

Isn't the code responsible for doing that inside `MiddlewareAdapter`; i.e.:
```js
class MiddlewareAdapter implements HydrogenKernelMiddleware {
// ...
  constructor(
    middleware: HydrogenKernelMiddleware,
    next: MiddlewareAdapter | KernelTransport
  ) {
    this._middleware = middleware;
    this._next = next;
}
// ...
  execute(
    code: string,
    callWatches: boolean,
    onResults: ResultsCallback
  ): void {

    for(let next = this._next; ; next = next._next) {

      if (next instaceof KernelTransport) {
        return next.execute(
          code,
          onResults
        );
      }

      if (next._middleware.execute) {
        return next._middleware.execute(
          next,
          code,
          onResults
        );
      }
    }
  }
// ...
}
```

> And for the plug in to have the option of skipping it's children (since now next.next is public API)?

I thought that was the case before (isn't this a consequence of the *precompiled* approach?).

> I think it's worth the extra code complexity in hydrogen to save plugins the hassle and plugin-side complexity of doing this. And to not open the possibility of plug-ins abusing this, e.g. by manually walking the next chain and modifying it.

Am I misunderstanding anything? How does a plugin transform the code passed on the next plugin?

---

Shall we take a break until tomorrow?

Reply:
(Please do respond on your own time)

In the code snippet you posted above, you have `class MiddlewareAdapter implements HydrogenKernelMiddleware` and `middleware: HydrogenKernelMiddleware`, i.e. both of them implement the same interface.

Your signature for `MiddlewareAdapter.execute` is `execute(code: string, callWatches: boolean, onResults: ResultsCallback) => void`. You also call `next._middleware.execute(next, code,onResults);`.

These two signatures are incompatible, so your proposed code above doesn't typecheck.

Reply:
> In the code snippet you posted above, you have class MiddlewareAdapter implements HydrogenKernelMiddleware and middleware: HydrogenKernelMiddleware, i.e. both of them implement the same interface.

yes, that's on purpose, so that an instance of `MiddlewareAdapter` invokes the next instance of `MiddlewareAdapter`.

> Your signature for MiddlewareAdapter.execute is execute(code: string, callWatches: boolean, onResults: ResultsCallback) => void. You also call next._middleware.execute(next, code, onResults);.
>
> These two signatures are incompatible (one starts with next, the other starts with code), so your proposed code above doesn't typecheck.

I've tried to fix it below by adding `next` to the signature of `execute`. Does it look OK?

```js
class MiddlewareAdapter implements HydrogenKernelMiddleware {
// ...
  constructor(
    middleware: HydrogenKernelMiddleware,
    next: MiddlewareAdapter | KernelTransport
  ) {
    this._middleware = middleware;
    this._next = next;
}
// ...
  execute(
    next: MiddlewareAdapter,
    code: string,
    callWatches: boolean,
    onResults: ResultsCallback
  ): void {

    for(let next = this._next; ; next = next._next) {

      if (next instaceof KernelTransport) {
        return next.execute(
          code,
          onResults
        );
      }

      if (next._middleware.execute) {
        return next._middleware.execute(
          next,
          code,
          onResults
        );
      }
    }
  }
// ...
}
```

Reply:
OK, now the code typechecks.

In the code, you call `next._middleware.execute(next, code,onResults);`, which is a method that a plugin defines. Consider how a plugin would implement a no-op execute method that passes through all requests.

Under your proposal above, that would look like:
```
kernel.addMiddleware({
  execute: (next, code, onResults) => {
    next.execute(next._next, code, onResults);
  }
})
```

Am I right?

Reply:
yes, I think that looks right.

Since the last plugin in the chain is `delegateToTransport`, all the other plugins only see `next._next` that are an instance of `MiddlewareAdapter`.

Reply:
Right, I see what you're suggesting. The change you proposed above definitely makes Hydrogen-internal code cleaner.

However, it also allows broken plugins like the following:

I.
```js
kernel.addMiddleware({
  execute: (next, code, onResults) => {
    next.execute(
      next._next, // My plugin doesn't like what comes next, so skip it!
      code, onResults);
  }
})
```

II.
```js
kernel.addMiddleware({
  execute: (next, code, onResults) => {
    next.execute(
      next, // Infinite loop, here we come!
      code, onResults);
  }
})
```

By introducing an extra interface within Atom, we permanently disallow plugins like I and II. I think it's worth it.

Reply:
> By introducing an extra interface within Hydrogen, we permanently disallow plugins like I and II. I think it's worth it.

Oh! now I understand the need for `HydrogenKernelMiddlewareStack`.

How about renaming the interface to `HydrogenKernelMiddlewareThunk` so that it's clear that this interface is just a wrapped version of `HydrogenKernelMiddleware`?

Reply:
That's a better name for it! Thanks for coming up with it.